### PR TITLE
Grab all No BP measures from the cache in one MGET

### PR DESCRIPTION
**Pull request by rsanheim**
**Story card:** [ch1646](https://app.clubhouse.io/simpledotorg/story/1646/report-perf-improvements)

### Before
Notice all the time spent in the green bar - those are many many GETs to redis for data.  For [one particular request in prod](https://app.datadoghq.com/apm/resource/simple_server/rack.request/4d88fc0b55eadeac?end=1602884493961&env=production&event=AQAAAXUzSIlcq_0HsQAAAABBWFV6U0pVaHRBSm9yamRkaDJVTA&graphType=flamegraph&index=apm-search&paused=false&query=env%3Aproduction%20service%3Asimple_server%20operation_name%3Arack.request%20resource_name%3A%22Reports%3A%3ARegionsController%23show%22&spanID=1013205513825382958&start=1602880893961&traceID=861081889945126204) I see over 3000 of them.

image

### CODE REVIEW

File: config/initializers/lograge.rb [hunk 1]
```diff
@@ -21,3 +22,13 @@ def fmt.call(data)
...
   end
 end
+
+ActiveSupport::Notifications.subscribe(/cache_read.*\.active_support/) do |name, start, finish, arg1, arg2|
+  RequestStore[:cache_stats] ||= {}
+  RequestStore[:cache_stats][:read] ||= 0
+  RequestStore[:cache_stats][:read] += 1
+end
+
+ActiveSupport::Notifications.subscribe(/cache_.*\.active_support/) do |name, start, finish, arg1, arg2|
```
Comment (vkrmis) [inline config/initializers/lograge.rb:32]: Is one of these `.subscribe` calls meant to be a `.instrument`? Or is that happening elsewhere?

File: app/services/no_bp_measure_service.rb [hunk 2]
```diff
@@ -14,20 +14,25 @@ def initialize(region, periods:)
...
-      result[period] = visited_without_bp_taken_count(period)
-    end
-  end
+  delegate :cache, to: Rails
 
-  def visited_without_bp_taken_count(period)
-    return 0 if facilities.empty?
-    Rails.cache.fetch(cache_key(period), version: cache_version, expires_in: CACHE_TTL, force: force_cache?) do
+  def call
+    cache_keys_to_period = periods.each_with_object({}) { |period, hsh|
```
Comment (vkrmis) [inline app/services/no_bp_measure_service.rb:20]: s/cache_keys_to_period/cache_keys_for_period ?

File: config/initializers/lograge.rb (hunk 1 above)
Comment (rsanheim) [reply config/initializers/lograge.rb]: Nope, these are subscribing to built in instrumentation [ActiveSupport](https://edgeguides.rubyonrails.org/active_support_instrumentation.html#active-support) has for caching.

File: app/services/no_bp_measure_service.rb [hunk 3]
```diff
@@ -1,33 +1,37 @@
...
-  CACHE_VERSION = 1
+  CACHE_VERSION = 2
   CACHE_TTL = 7.days
 
   def initialize(region, periods:)
     @region = region
     @periods = periods
     @facilities = region.facilities.to_a
     @facility_ids = @facilities.map(&:id)
+    @cache_keys_for_period = periods.each_with_object({}) { |period, hsh| hsh[cache_key(period)] = period }
```
Comment (timcheadle) [inline app/services/no_bp_measure_service.rb:10]: Feels like this should be a memoized method instead

**Review by timcheadle — APPROVED** [review]
Approved, had a style suggestion

**Pull request merged**